### PR TITLE
Feature: Sftp Provisioner and Client

### PR DESCRIPTION
**Pull request by bcronin90**
Add a client and provisioner for SFTP transfers. Provisioner is NoOp, so it generates users and locations from data addresses, but doesn't provision objects cloud-side.

Signed-off-by: Brendan Cronin [brendan.cronin@mercedes-benz.com](mailto:brendan.cronin@mercedes-benz.com)

### CODE REVIEW

File: edc-extensions/pom.xml [hunk 1]
```diff
@@ -42,7 +42,9 @@
         <module>dataplane-selector-configuration</module>
         <module>data-encryption</module>
         <module>cx-oauth2</module>
-        <module>control-plane-adapter</module>
```
Comment (denisneuling) [inline edc-extensions/pom.xml:45]: This line perhaps has been removed accidentially - let's add it again then the build should be back to green

Comment (denisneuling) [reply edc-extensions/pom.xml:45]: Nevermind, looks like that module is not present on your branch. Will submit an updating commit.

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpDataSinkFactory.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    if (request.getDestinationDataAddress().getType().equals("sftp")
+        && request.getDestinationDataAddress() instanceof SftpDataAddress) {
+      return VALID;
+    }
+    return Result.failure(String.format("Invalid dataflow request: %s", request.getId()));
+  }
+
+  @Override
+  public DataSink createSink(DataFlowRequest request) {
+    SftpDataAddress destination = (SftpDataAddress) request.getDestinationDataAddress();
```
Comment (ndr-brt) [inline edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpDataSinkFactory.java:51]: This will throw exception, you'll need to build a new `SftpDataAddress` with the properties, take a look at `HttpRequestParamsSupplier` for an example

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpDataSourceFactory.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    if (request.getSourceDataAddress().getType().equals("sftp")
+        && request.getSourceDataAddress() instanceof SftpDataAddress) {
+      return VALID;
+    }
+    return Result.failure(String.format("Invalid dataflow request: %s", request.getId()));
+  }
+
+  @Override
+  public DataSource createSource(DataFlowRequest request) {
+    SftpDataAddress source = (SftpDataAddress) request.getSourceDataAddress();
```
Comment (ndr-brt) [inline edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpDataSourceFactory.java:51]: This will throw an exception, like above

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpDataSinkFactory.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+@RequiredArgsConstructor
+public class SftpDataSinkFactory implements DataSinkFactory {
+
+  @NotNull SftpClientWrapper sftpClientWrapper;
+
+  @Override
+  public boolean canHandle(DataFlowRequest request) {
+    return request.getDestinationDataAddress().getType().equals("sftp")
+        && request.getDestinationDataAddress() instanceof SftpDataAddress;
```
Comment (ndr-brt) [inline edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpDataSinkFactory.java:37]: the `instanceof` condition is always false (same on `validate` and on the SourceFactory class

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  @Setter private int connectionTimeoutSeconds = DEFAULT_CONNECTION_TIMEOUT_SECONDS;
+
+  @Override
+  public void uploadFile(
+      @NonNull final SftpUser sftpUser,
+      @NonNull final SftpLocation sftpLocation,
+      @NonNull final InputStream inputStream,
+      @NonNull Collection<SftpClient.OpenMode> openModes)
+      throws IOException {
+    try (final SftpClient sftpClient = getSftpClient(sftpUser, sftpLocation)) {
```
Comment (ndr-brt) [inline edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java:62]: seems to be a little expensive to have to create a new client and a new session for every operation. I would follow the typical pattern where the `Factory` creates/obtain an already existing client and then it passes it to the `Source`/`Sink` (take a look at the s3 or gcs extensions for examples)

**Review by ndr-brt — COMMENTED** [review]
Great! 
I would suggest to add an e2e test that actually loads the provision and the data-plane extensions and do a real e2e transfer, because looks that there are some integration things to be fixed here and there, and the test would definitely help out.

File: edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import org.jetbrains.annotations.Nullable;
+
+@RequiredArgsConstructor
+public class SftpProviderResourceDefinitionGenerator
+    implements ProviderResourceDefinitionGenerator {
+
+  @Override
+  public @Nullable ResourceDefinition generate(
+      DataRequest dataRequest, DataAddress assetAddress, Policy policy) {
+    if (!(assetAddress instanceof SftpDataAddress)
```
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java]: Would I then just try to parse the `connectorAddress` of the `dataRequest`? Because I'm not sure where else to get the connection parameters.

File: edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        || !assetAddress.getType().equals(DATA_ADDRESS_TYPE)) {
+      throw new EdcSftpException(
+          String.format("Data Address %s is not an SFTP Data Address", assetAddress.getKeyName()));
+    }
+    SftpDataAddress sftpDataAddress = (SftpDataAddress) assetAddress;
```
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java:40]: The problem is that the default `DataAddress` class contains only a vague field called "properties" and I basically have to hope that whatever extension created this particular address conforms to some unwritten agreement to include the relevant connection info. I'm really not sure how I would parse that into something usable.

File: edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/NoOpSftpLocationFactory.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org.eclipse.tractusx.edc.transferprocess.sftp.provisioner;
+
+import lombok.RequiredArgsConstructor;
+import org.eclipse.tractusx.edc.transferprocess.sftp.common.SftpLocation;
+import org.eclipse.tractusx.edc.transferprocess.sftp.common.SftpLocationFactory;
+
+@RequiredArgsConstructor
+public class NoOpSftpLocationFactory implements SftpLocationFactory {
```
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/NoOpSftpLocationFactory.java:22]: Because they don't actually create any resources cloud-side. They assume that users and permissions already exist and only create the objects necessary for the client to interact with it. Is there a better way to express this in a name?

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java (hunk 5 above)
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java:63]: The issue there is that a `ClientSession` is specific to a certain location and an `SshClient`, which underlies the `SftpClient` is specific to the user. And both are dependent on the data address. So we need a new client each time.

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientExtension.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+public class SftpClientExtension implements ServiceExtension {
+  @Override
+  public void initialize(ServiceExtensionContext context) {
+    SftpClientWrapperImpl sftpClient = new SftpClientWrapperImpl();
+    SftpDataSinkFactory sftpDataSinkFactory = new SftpDataSinkFactory(sftpClient);
+    SftpDataSourceFactory sftpDataSourceFactory = new SftpDataSourceFactory(sftpClient);
+
+    context.registerService(SftpClientWrapper.class, sftpClient);
+    context.registerService(DataSinkFactory.class, sftpDataSinkFactory);
+    context.registerService(DataSourceFactory.class, sftpDataSourceFactory);
```
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientExtension.java]: That gives me a dependency error, even when importing the `data-plane-spi`. Do I need any further modules?

File: edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ndr-brt) [reply edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java]: no, it's just enough to leave the second part of the condition: `!assetAddress.getType().equals(DATA_ADDRESS_TYPE)`

Maybe this shouldn't throw an exception but return null (according to `ProviderResourceDefinitionGenerator` documentation

File: edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/NoOpSftpLocationFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ndr-brt) [reply edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/NoOpSftpLocationFactory.java:22]: I would just call it `SftpLocationFactoryImpl`, it doesn't give the impression to create resources, just a `SftpLocation` object instance

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java (hunk 5 above)
Comment (ndr-brt) [reply edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java:63]: Yeah, I put this in the wrong way.
Why don't create the client in the `Sink`/`Source` `Factory` from the credentials and include it in the `DataSource`/`DataSink`. 
It would be more coherent with the other data-plane extensions.

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientExtension.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ndr-brt) [reply edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientExtension.java]: The `PipelineService` is provided by `data-plane-core`, that dependency should be already available in our `data-plane` build dependency list, if you need that for testing purposes, just add it with `test` scope: `data-plane-core`

File: edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/SftpProviderResourceDefinitionGenerator.java]: I went for a slightly more complicated solution, have a look if you like it. Basically I really don't want to just dump everything into an undocumented properties field.

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientExtension.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientExtension.java]: Thanks. We were missing that dependency all together.

File: edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java (hunk 5 above)
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java:63]: And then we wouldn't register `SftpClientWrapper.class` at all? Only the sink/source?

Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-client/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/client/SftpClientWrapperImpl.java:63]: I did exactly that. Lot of refactoring to get there, but it's done.

File: edc-extensions/transferprocess-sftp-common/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/common/SftpDataAddress.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  private static final String LOCATION_PATH = "locationPath";
+  private static final String USER_NAME = "userName";
+  private static final String USER_PASSWORD = "userPassword";
+  private static final String USER_PRIVATE_KEY = "userPrivateKey";
+
+  @Getter private static final String CONNECTION_TYPE = "sftp";
+  @NonNull private final SftpUser sftpUser;
+  @NonNull private final SftpLocation sftpLocation;
+
+  public static SftpDataAddress fromDataAddress(DataAddress dataAddress) {
```
Comment (ndr-brt) [inline edc-extensions/transferprocess-sftp-common/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/common/SftpDataAddress.java:25]: The `type` property on the input `dataAddress` should be checked as well. 
maybe some tests covering this method would help out handling some edge cases, because seems to be a piece of code that could throw unexpected exceptions.
The whole method could be wrapped into a `try-catch` block, returning a `Result<SftpDataAddress>`, so no caller will have to catch unchecked exceptions, just to verify if the `Result` is succeeded or not.

Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-common/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/common/SftpDataAddress.java:25]: Good points on the Exception handling. I have an issue with the `type` that is just another value floating around in the property soup - it could be anything. I have no idea what I would even check for.

Comment (ndr-brt) [reply edc-extensions/transferprocess-sftp-common/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/common/SftpDataAddress.java:25]: it could be anything, but in this case it has to be `sftp`, so it must be checked, otherwise the `DataAddress` could not be converted to an `SftpDataAddress`.

File: edc-extensions/transferprocess-sftp-client/pom.xml [hunk 11]
```diff
@@ -129,6 +129,10 @@
             <groupId>org.eclipse.dataspaceconnector</groupId>
             <artifactId>data-plane-spi</artifactId>
         </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-client/pom.xml:132]: Not sure how I would move it there. Isn't that upstream?

File: edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/NoOpSftpProvider.java [hunk 12]
```diff
@@ -20,14 +20,26 @@
 
 public class NoOpSftpProvider implements SftpProvider {
   @Override
-  public void createUser(SftpUser user) {}
+  public void createUser(SftpUser user) {
+    // This provisioner does not create cloud resources.
```
[the remark this answers is not in the document]
Comment (bcronin90) [reply edc-extensions/transferprocess-sftp-provisioner/src/main/java/org/eclipse/tractusx/edc/transferprocess/sftp/provisioner/NoOpSftpProvider.java]: Good catch.

**Review by ndr-brt — APPROVED** [review]
